### PR TITLE
fuzz: bolster ExtractDestination(s) checks

### PR DESCRIPTION
**Pull request by mjdietzx**

### CODE REVIEW

File: src/test/fuzz/script.cpp [hunk 1]
```diff
@@ -56,22 +56,43 @@ FUZZ_TARGET_INIT(script, initialize_script)
...
     int required_ret;
-    (void)ExtractDestinations(script, type_ret, addresses, required_ret);
-
-    const FlatSigningProvider signing_provider;
-    (void)InferDescriptor(script, signing_provider);
-
-    (void)IsSegWitOutput(signing_provider, script);
-
-    (void)IsSolvable(signing_provider, script);
+    if (!ExtractDestinations(script, type_ret, addresses, required_ret)) {
```
Comment (sipa) [inline src/test/fuzz/script.cpp:62]: This code could be significantly more efficient by just calling `ExtractDestination`(`s`) once each, caching the result, and then making assertions on those results.

**Comment by sipa** [comment]
Concept ACK, but probably not worth spending too much time on if `ExtractDestinations` is going away anyway.

**Comment by practicalswift** [comment]
Concept ACK

Thanks for improving the fuzzing harnesses!

**Comment by mjdietzx** [comment]
rebased to master - fixed conflicts after https://github.com/bitcoin/bitcoin/pull/20765 was merged

File: src/test/fuzz/script.cpp [hunk 2]
```diff
@@ -56,22 +56,45 @@ FUZZ_TARGET_INIT(script, initialize_script)
...
+        if (type_ret == TxoutType::MULTISIG) {
+            assert(addresses.empty() && required_ret == 0);
+        } else {
+            assert(type_ret == TxoutType::PUBKEY ||
+                   type_ret == TxoutType::NONSTANDARD ||
+                   type_ret == TxoutType::NULL_DATA);
+        }
+    } else {
+        assert(required_ret >= 1 && required_ret <= 16);
+        assert((unsigned long)required_ret == addresses.size());
```
Comment (practicalswift) [inline src/test/fuzz/script.cpp:75]: Nit: `size_t` instead of `unsigned long` to match the return type of `size()`?

**Review by practicalswift — COMMENTED** [review]
Tested ACK a29f522ba4aa71582b54025c5682b4c1687ae9f3

Thanks for adding assertions to existing fuzzing harnesses: that's a great way to improve the value provided by them!

Feel free to ignore nit.

**Comment by adamjonas** [comment]
Tested a29f522ba4aa71582b54025c5682b4c1687ae9f3 with `FUZZ=script src/test/fuzz/fuzz qa-assets/fuzz_seed_corpus/script`. No problems.

**Pull request merged**